src/traits.rs

```diff
@@ -871,6 +939,22 @@ macro_rules! prim_int_impl {
                 <$T>::rotate_right(self, n)
             }
 
+            fn arithmetic_shl(self, n: u32) -> Self {
+                ((self as $S) << n) as $T
```

Comment:
What exactly are you asking about?

Note that shl behaves the same either way, logical or arithmetic.  For shr, the [reference](http://doc.rust-lang.org/reference.html#bitwise-operators) calls it an arithmetic shift, but that doesn't actually do sign-extension for unsigned types -- much like C.